### ISSUE TITLE
This needs to work on older phones....

### ISSUE THREAD

**Issue by frohro** [pedronveloso/android-digital-frame#2]
I can run it on Android 10, but not 7 and there are plenty of phones that run Android 4.  If this is going to be good for recycling old phones it needs to run on older phones.
Thanks for all your work on this!

**Comment by frohro** [comment]
o@pedronveloso Another thing I am fin.  ding.  The weather is not accurate.  I think this is because it is not updated very often.  I suggest hourly.  The cost of the API if you are just doing weather as opposed to forecast is free.

**Comment by pedronveloso** [comment]
@frohro agreed on the backwards support. I'll see how far back I can easily backport it.

Regarding weather, it is currently using the weather API for `current weather` and it refreshes hourly. However even this one is only free up to a certain amount of calls a day ( https://openweathermap.org/price ) - not an issue right now, but will be once I publish on the Play Store and more people start using it. Caching this is not really an feasable option, because there are more locations around the world than there are available API calls in a day, however I've contact their support to possibly get a key that would allow for this simple usecase with higher limits - since its an OSS project. One possible reason why you might be seeing the weather prection off might have to do with fast changing weather temperatures - in many places at sunrise or sunset the temperature flactuates a lot even within the hour. So I'll likely have to use daily forecast let's say every 6h instead of asking every hour, which should give us better accuracy and efficiency. More to come on this.

**Comment by pedronveloso** [comment]
I've backported the app all the way to API Level 21, that's Android 5. According to [apilevels](https://apilevels.com/) that represents 99.5% of the devices out there (although that number is probably based on active devices). 

This is the minimum supported version for several of the project dependencies, so that's somewhat of a hard limit without compromising the usage of a fairly modern tech stack. I feel good about this.

I've also improved the fullscreen support as well, let me know if you still see the nav bar on your device.

**Comment by frohro** [comment]
I haven't yet tried it on an older phone, but the navigation bars behave the same as before.

**Comment by frohro** [comment]
So I checked my LG V20 running Android 8.0.  I get this:
 E  Well this is awkward; package com.pedronveloso.digitalframe.DigitalFrameApplication had UID -1
                                                                                                    java.lang.Throwable
                                                                                                    	at com.android.server.pm.PackageDexOptimizer.performDexOptLI(PackageDexOptimizer.java:148)
                                                                                                    	at com.android.server.pm.PackageDexOptimizer.performDexOpt(PackageDexOptimizer.java:123)
                                                                                                    	at com.android.server.pm.PackageManagerService.installPackageLI(PackageManagerService.java:19431)
                                                                                                    	at com.android.server.pm.PackageManagerService.installPackageTracedLI(PackageManagerService.java:18976)
                                                                                                    	at com.android.server.pm.PackageManagerService.-wrap35(Unknown Source:0)
                                                                                                    	at com.android.server.pm.PackageManagerService$6.run(PackageManagerService.java:16235)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:789)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:98)
                                                                                                    	at android.os.Looper.loop(Looper.java:164)
                                                                                                    	at android.os.HandlerThread.run(HandlerThread.java:65)
                                                                                                    	at com.android.server.ServiceThread.run(ServiceThread.java:46)

**Comment by pedronveloso** [comment]
> I haven't yet tried it on an older phone, but the navigation bars behave the same as before.

Can you share a screenshot with me, so I can have a better idea of what it looks like? And what device is this happening (model/make and Android version are enough).

The errors on your LG V20 is during installation, or runtime ?

**Comment by frohro** [comment]
Here is a screenshot.  https://photos.app.goo.gl/NxirYqtGTurSXEeH9 
This is on an original Pixel running Android 10.
The error in the LG V20 running Android 8 seems to be when it gets loaded on the phone over the USB debugging, so I'm not sure.   This particular LG was rooted.  I didn't fiddle with it a long time.  I have another one at school, and I'll bring it back home tomorrow.  That one is not rooted and may have a different Android version on it.  Is there a way I can make an APK with Android Studio and transfer it some other way and try running it that way?  I haven't played with Android Studio since Android 4, and a lot has changed.  :-)

**Comment by frohro** [comment]
So I tried it on another LG V20 (running Android 7.0).  It crashes after loading the program with this message when I open the settings.
FATAL EXCEPTION: main
                                                                                                    Process: com.pedronveloso.digitalframe, PID: 11149
                                                                                                    java.lang.OutOfMemoryError: Failed to allocate a 315607052 byte allocation with 17212554 free bytes and 231MB until OOM
                                                                                                    	at dalvik.system.VMRuntime.newNonMovableArray(Native Method)
                                                                                                    	at android.graphics.BitmapFactory.nativeDecodeAsset(Native Method)
                                                                                                    	at android.graphics.BitmapFactory.decodeStream(BitmapFactory.java:620)
                                                                                                    	at android.graphics.BitmapFactory.decodeResourceStream(BitmapFactory.java:455)
                                                                                                    	at android.graphics.drawable.Drawable.createFromResourceStream(Drawable.java:1152)
                                                                                                    	at android.content.res.ResourcesImpl.loadDrawableForCookie(ResourcesImpl.java:855)
                                                                                                    	at android.content.res.ResourcesImpl.loadDrawable(ResourcesImpl.java:706)
                                                                                                    	at android.content.res.Resources.getDrawable(Resources.java:776)
                                                                                                    	at android.content.Context.getDrawable(Context.java:525)
                                                                                                    	at com.pedronveloso.digitalframe.elements.background.PhotosBackgroundViewModel$calculateBrightness$1.invokeSuspend(PhotosBackgroundViewModel.kt:93)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:367)
                                                                                                    	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:30)
                                                                                                    	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:25)
                                                                                                    	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:110)
                                                                                                    	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)
                                                                                                    	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:56)
                                                                                                    	at kotlinx.coroutines.BuildersKt.launch(Unknown Source)
                                                                                                    	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
                                                                                                    	at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source)
                                                                                                    	at com.pedronveloso.digitalframe.elements.background.PhotosBackgroundViewModel.calculateBrightness(PhotosBackgroundViewModel.kt:89)
                                                                                                    	at com.pedronveloso.digitalframe.elements.background.PhotosBackgroundViewModel.backgroundRotation(PhotosBackgroundViewModel.kt:81)
                                                                                                    	at com.pedronveloso.digitalframe.elements.background.PhotosBackgroundViewModel.access$backgroundRotation(PhotosBackgroundViewModel.kt:48)
                                                                                                    	at com.pedronveloso.digitalframe.elements.background.PhotosBackgroundViewModel$backgroundRotation$1.invokeSuspend(PhotosBackgroundViewModel.kt)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTaskKt.resume(DispatchedTask.kt:235)
                                                                                                    	at kotlinx.coroutines.DispatchedTaskKt.dispatch(DispatchedTask.kt:168)
                                                                                                    	at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:474)
                                                                                                    	at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:508)
                                                                                                    	at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl$default(CancellableContinuationImpl.kt:497)
                                                                                                    	at kotlinx.coroutines.CancellableContinuationImpl.resumeUndispatched(CancellableContinuationImpl.kt:595)
                                                                                                    	at kotlinx.coroutines.android.HandlerContext$scheduleResumeAfterDelay$$inlined$Runnable$1.run(Runnable.kt:19)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:751)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                                                    	at android.os.Looper.loop(Looper.java:154)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:6247)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:872)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:762)
                                                                                                    	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@dd440b, Dispatchers.Main.immediate]

**Comment by pedronveloso** [comment]
@frohro I think I figured out the issue there. By any chance did you have the same google account on different devices? I noticed that the photos were being backed-up using Google's built-in mechanism. This is a nice to have for app settings, but not so much for photos since when loading them the app will resize them accordingly to the screen size. So a device with a big resolution would make an older and smaller resolution device run out of memory easily. I recommend getting the latest changes, and completely removing the app (or clearing data) before reinstalling.

I've also attempted something else to force the full-screen. If it still doesn't work, might require some trial and error. I can't repro the issue on my end, might be some system setting that I'm unfamiliar with.

**Comment by frohro** [comment]
@pedronveloso , good work!  I am using the same Google account on all the devices.  I think you have fixed the navigation button (full screen) issue!   On the Pixel 1, I am not getting weather data now.  The LG V20 with Android 7 doesn't seem to crash as quickly, but it does sometimes.  I had some difficulty getting it to change photos from default, but it eventually worked.  Changing again I had the same issues.  It seems you have to make sure the app quits before the photos will change.  You need to check and make sure the latitude and longitude make sense.  I neglected to put in the decimal point on the Pixel and that was the issue with the weather.   It too needed to be force-quit and restarted to get the new setting to be read.

**Comment by frohro** [comment]
@pedronveloso I think the other LG V20 that is rooted may have a hardware problem with the USB connection.  It keeps telling me it is connected, but offline and the USB adb connection seems a bit flakey.

**Comment by pedronveloso** [comment]
For the LG that has trouble via USB, I recommend generating the APK and sending it to yourself uploading it somewhere. In Android studio do this to build the APK and get access to the file: https://www.geeksforgeeks.org/how-to-generate-unsigned-shareable-apk-in-android-studio/

Sometime soon I plan to have the app on the play store, which should make it easier.

Yeah, there's no proper photo management yet. One has to clear app data and import photos all over again. I need to look to why it is still crashing. It didn't recover photos from your Google account this time correct?

Good point regarding coordinate field validation. I'll add that.

**Comment by pedronveloso** [comment]
And it is correct that sometimes settings don't immediately take effect. This is on my to-do list. The fact that they sometimes work at all is just a happy side effect, I haven't coded it purposely that way yet :)

**Comment by frohro** [comment]
@pedronveloso, there needs to be a way to reset the photos without resetting all the other settings.

**Comment by frohro** [comment]
@pedronveloso, one more thing I noticed is that the settings do not respond to the "Done" button on the check mark on the keyboard.  You have to post the navigate back button for the whole device.

**Comment by frohro** [comment]
I got it to work on my rooted LG V20 with Android 8, and a flakey USB connection.

**Comment by frohro** [comment]
I would wait on the Play Store until you are really certain the bugs are fixed.  Reviewers can be merciless, and once you get a bad reputation, digging out of it is really hard, because the few number of stars discourages others from even trying it.  There is a great "market" for it though, because I tried all the other ones, and this one has the most promise so far.

**Comment by pedronveloso** [comment]
Just posting an update here, stating that version `0.0.8` does some minor improvements on topics mentioned on this thread, and it warrants a full reinstall on devices that were having issues running the app.

That being said, a few issues are still pending, namely:
 - Proper testing and tuning for low-end and/or  older devices
 - Settings changes should take effect right away (plan to work on this next)

**Comment by pedronveloso** [comment]
I'm closing this issue, but this hasn't been forgotten. I'm closing the GH Issues functionality on the project, and opening GH Discussions as it has more value IMO
